unity/Holo/Assets/Scripts/BlendShapeAnimation.cs

```diff
@@ -1,76 +1,101 @@
-
-        skinnedMeshRenderer.SetBlendShapeWeight(currentIndex, 100f);
-        skinnedMeshRenderer.SetBlendShapeWeight(zeroIndex, 0.0f);
-    }
-
-    void UpdateCyclic()
-    {
-        int zeroIndex = currentIndex;
-        currentIndex = (currentIndex + 1) % blendShapeCount;
-
-        skinnedMeshRenderer.SetBlendShapeWeight(currentIndex, 100f);
-        skinnedMeshRenderer.SetBlendShapeWeight(zeroIndex, 0.0f);
-    }
-
-    void UpadteCyclicWithLastZero()
-    {
-        int cycleCount = blendShapeCount + 1;
-        int zeroIndex = currentIndex;
-
-        currentIndex = (currentIndex + 1) % cycleCount;
-        if(currentIndex < blendShapeCount)
-            skinnedMeshRenderer.SetBlendShapeWeight(currentIndex, 100f);
-        if(zeroIndex < blendShapeCount)
-            skinnedMeshRenderer.SetBlendShapeWeight(zeroIndex, 0.0f);
-       
-    }
-}
\ No newline at end of file
+﻿//Using C#
+
+using UnityEngine;
+using System.Collections;
+
+public class BlendShapeAnimation : MonoBehaviour
+{
+    int blendShapeCount;
+    SkinnedMeshRenderer skinnedMeshRenderer;
+    Mesh skinnedMesh;
+
+    float currentIndex = 0f;
+
+    public bool mirrorAnimation = false;
+    public float speed = 1f;
+
+    void Awake()
+    {
+        skinnedMeshRenderer = GetComponent<SkinnedMeshRenderer>();
+        skinnedMesh = GetComponent<SkinnedMeshRenderer>().sharedMesh;
+    }
+
+    void Start()
+    {
+        blendShapeCount = skinnedMesh.blendShapeCount;
+        Debug.Log("BlendShapes count: " + blendShapeCount);
+    }
+
+    private void Update()
+    {
+        if (mirrorAnimation) UpdateMirror();
+        else UpdateCyclic();
+    }
+
+    private bool mirrorIncreasing;
+
+    void UpdateMirror()
+    {
+        float maxCurrentIndex = blendShapeCount - 1f;
+        if (mirrorIncreasing)
+        {
+            currentIndex += Time.deltaTime * speed;
+            if (currentIndex > maxCurrentIndex)
+            {
+                mirrorIncreasing = false;
+                currentIndex = maxCurrentIndex - (currentIndex - maxCurrentIndex);
+            }
+        }
+        else
+        {
+            currentIndex -= Time.deltaTime * speed;
+            if (currentIndex < 0f)
+            {
+                mirrorIncreasing = true;
+                currentIndex = -currentIndex;
+            }
+
+        }
+
+        UpdateBlendShapes();
+    }
+
+    void UpdateCyclic()
+    {
+        currentIndex += Time.deltaTime * speed;
+        currentIndex = Mathf.Repeat(currentIndex, blendShapeCount);
+
+        UpdateBlendShapes();
+    }
+
+    /* Update current mesh shape, looking at currentIndex. */
+    void UpdateBlendShapes()
+    {
+        int previousShape = (int)currentIndex;
+        float frac = currentIndex - previousShape;
+        int nextShape = previousShape + 1;
+        if (nextShape > blendShapeCount - 1)
+        {
+            nextShape = 0;
+        }
+
+        for (int i = 0; i < blendShapeCount; i++)

```

Comment:
Do we really need to iterate over all BlendShapes each time?

Reply:
Good point, improved:)